core/sql/exp/exp_datetime.cpp

```diff
@@ -2410,773 +2550,363 @@ ExpDatetime::convAsciiToDate(char *srcData,
-	    //
-	    ExRaiseSqlError(heap, diagsArea, EXE_CONVERT_STRING_ERROR);
-	    ExRaiseSqlError(heap, diagsArea, EXE_CONVERT_DATETIME_ERROR);
-	    return -1;
-	  }
-	  
-	  srcData += 2;
+      // the second
+      if (convSrcDataToDst(2, srcData, 1, &dstData[6], NULL, heap, diagsArea))
+        return -1;
 
-	  if (*srcData != ':') {
-	    // string contains invalid delimiter
-	    //
-	    ExRaiseSqlError(heap, diagsArea, EXE_CONVERT_DATETIME_ERROR);
-	    return -1;
-	  }
-	  srcData++;  // move on to the next field
-	  
-	  // the minute
-	  if (isDigit8859_1(*srcData) && isDigit8859_1(*(srcData+1))) {
-#pragma nowarn(1506)   // warning elimination 
-	    dstData[5] = char (*srcData - '0') * 10 + (*(srcData+1) - '0');
-#pragma warn(1506)  // warning elimination 
-	  } else {
-	    // string contains non-digit charecter(s)
-	    //
-	    ExRaiseSqlError(heap, diagsArea, EXE_CONVERT_STRING_ERROR);
-	    ExRaiseSqlError(heap, diagsArea, EXE_CONVERT_DATETIME_ERROR);
-	    return -1;
-	  }
-	  
-	  srcData += 2;
-	  if (*srcData != ':') {
-	    // string contains invalid delimiter
-	    //
-	    ExRaiseSqlError(heap, diagsArea, EXE_CONVERT_DATETIME_ERROR);
-	    return -1;
-	  }
-	  srcData++;  // move on to the next field
-	  
-	  // the second
-	  if (isDigit8859_1(*srcData) && isDigit8859_1(*(srcData+1))) {
-#pragma nowarn(1506)   // warning elimination 
-	    dstData[6] = char (*srcData - '0') * 10 + (*(srcData+1) - '0');
-#pragma warn(1506)  // warning elimination 
-	  } else {
-	    // string contains non-digit charecter(s)
-	    //
-	    ExRaiseSqlError(heap, diagsArea, EXE_CONVERT_STRING_ERROR);
-	    ExRaiseSqlError(heap, diagsArea, EXE_CONVERT_DATETIME_ERROR);
-	    return -1;
-	  }
-	  
-	  srcData += 2;
-	  
-	  dstData[7]  = 0;
-	  dstData[8]  = 0;
-	  dstData[9]  = 0;
-	  dstData[10] = 0;
-	}
-    };  // case USA format
+      dstData[7]  = 0;
+      dstData[8]  = 0;
+      dstData[9]  = 0;
+      dstData[10] = 0;
+     };
     break;
 
-  case DATETIME_FORMAT_USA3:
-  case DATETIME_FORMAT_USA4:
-  case DATETIME_FORMAT_USA5:
-  case DATETIME_FORMAT_USA8:
+  case DATETIME_FORMAT_USA3: // YYYY/MM/DD
+  case DATETIME_FORMAT_USA4: // YYYYMMDD
+  case DATETIME_FORMAT_USA5: // YY/MM/DD
     {
-      // this is USA format: See exp_datetime.h for details.
+      // the year
+      Lng32 numYearDigits = (srcFormat == DATETIME_FORMAT_USA5 ? 2 : 4);
+      char sep = '/';
+      char * septr = (srcFormat == DATETIME_FORMAT_USA4 ? NULL : &sep);
 
       // the year
-      Lng32 numYearDigits = 4;
-      if (srcFormat == DATETIME_FORMAT_USA5)
-	numYearDigits = 2;
-      for (year = 0, i = 0; 
-	   i < numYearDigits && isDigit8859_1(*srcData); i++, srcData++)
-#pragma nowarn(1506)   // warning elimination 
-        year = year * 10 + *srcData - '0';
-#pragma warn(1506)  // warning elimination 
+      if (convSrcDataToDst(numYearDigits, srcData, 2, dstData, septr, heap, diagsArea))
+        return -1;
```

Comment:
Do we need to add 2000 to (short)dstData when it is a YY format (as opposed to a YYYY format)?